File: tmp_cfr/decompiled/org/cobalt/api/module/setting/impl/ColorSetting.java
/*
 * Decompiled with CFR 0.152.
 * 
 * Could not load the following classes:
 *  com.google.gson.JsonElement
 *  com.google.gson.JsonObject
 *  com.google.gson.JsonPrimitive
 *  kotlin.Metadata
 *  kotlin.NoWhenBranchMatchedException
 *  kotlin.jvm.internal.Intrinsics
 *  kotlin.ranges.RangesKt
 *  org.jetbrains.annotations.NotNull
 */
package org.cobalt.api.module.setting.impl;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;
import java.awt.Color;
import kotlin.Metadata;
import kotlin.NoWhenBranchMatchedException;
import kotlin.jvm.internal.Intrinsics;
import kotlin.ranges.RangesKt;
import org.cobalt.api.module.setting.Setting;
import org.cobalt.api.module.setting.impl.ColorMode;
import org.cobalt.api.module.setting.impl.ThemeColorResolver;
import org.cobalt.api.ui.theme.Theme;
import org.cobalt.api.ui.theme.ThemeManager;
import org.jetbrains.annotations.NotNull;

@Metadata(mv={2, 3, 0}, k=1, xi=48, d1={"\u0000:\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\u0010\b\n\u0002\u0010\u000e\n\u0002\b\b\n\u0002\u0010\u0007\n\u0002\b\u0006\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u0002\n\u0002\b\u0007\n\u0002\u0018\u0002\n\u0002\b\u0006\n\u0002\u0010\t\n\u0002\b\b\u0018\u00002\b\u0012\u0004\u0012\u00020\u00020\u0001B\u001f\u0012\u0006\u0010\u0004\u001a\u00020\u0003\u0012\u0006\u0010\u0005\u001a\u00020\u0003\u0012\u0006\u0010\u0006\u001a\u00020\u0002\u00a2\u0006\u0004\b\u0007\u0010\bJ\u000f\u0010\t\u001a\u00020\u0002H\u0002\u00a2\u0006\u0004\b\t\u0010\nJ7\u0010\u0011\u001a\u00020\u00022\u0006\u0010\u000b\u001a\u00020\u00022\u0006\u0010\r\u001a\u00020\f2\u0006\u0010\u000e\u001a\u00020\f2\u0006\u0010\u000f\u001a\u00020\f2\u0006\u0010\u0010\u001a\u00020\fH\u0002\u00a2\u0006\u0004\b\u0011\u0010\u0012J\u0017\u0010\u0016\u001a\u00020\u00152\u0006\u0010\u0014\u001a\u00020\u0013H\u0016\u00a2\u0006\u0004\b\u0016\u0010\u0017J\u000f\u0010\u0018\u001a\u00020\u0013H\u0016\u00a2\u0006\u0004\b\u0018\u0010\u0019R\u001a\u0010\u0006\u001a\u00020\u00028\u0016X\u0096\u0004\u00a2\u0006\f\n\u0004\b\u0006\u0010\u001a\u001a\u0004\b\u001b\u0010\u001cR\"\u0010\u001e\u001a\u00020\u001d8\u0006@\u0006X\u0086\u000e\u00a2\u0006\u0012\n\u0004\b\u001e\u0010\u001f\u001a\u0004\b \u0010!\"\u0004\b\"\u0010#R\u0014\u0010%\u001a\u00020$8\u0002X\u0082\u0004\u00a2\u0006\u0006\n\u0004\b%\u0010&R$\u0010+\u001a\u00020\u00022\u0006\u0010'\u001a\u00020\u00028V@VX\u0096\u000e\u00a2\u0006\f\u001a\u0004\b(\u0010\u001c\"\u0004\b)\u0010*\u00a8\u0006,"}, d2={"Lorg/cobalt/api/module/setting/impl/ColorSetting;", "Lorg/cobalt/api/module/setting/Setting;", "", "", "name", "description", "defaultValue", "<init>", "(Ljava/lang/String;Ljava/lang/String;I)V", "resolveColor", "()I", "argb", "", "hueShift", "saturationMult", "brightnessMult", "opacityMult", "tweakColor", "(IFFFF)I", "Lcom/google/gson/JsonElement;", "element", "", "read", "(Lcom/google/gson/JsonElement;)V", "write", "()Lcom/google/gson/JsonElement;", "I", "getDefaultValue", "()Ljava/lang/Integer;", "Lorg/cobalt/api/module/setting/impl/ColorMode;", "mode", "Lorg/cobalt/api/module/setting/impl/ColorMode;", "getMode", "()Lorg/cobalt/api/module/setting/impl/ColorMode;", "setMode", "(Lorg/cobalt/api/module/setting/impl/ColorMode;)V", "", "instanceStartTime", "J", "newValue", "getValue", "setValue", "(I)V", "value", "cobalt"})
public final class ColorSetting
extends Setting<Integer> {
    private final int defaultValue;
    @NotNull
    private ColorMode mode;
    private final long instanceStartTime;

    public ColorSetting(@NotNull String name, @NotNull String description, int defaultValue) {
        Intrinsics.checkNotNullParameter((Object)name, (String)"name");
        Intrinsics.checkNotNullParameter((Object)description, (String)"description");
        super(name, description, defaultValue);
        this.defaultValue = defaultValue;
        this.mode = new ColorMode.Static(defaultValue);
        this.instanceStartTime = System.currentTimeMillis();
    }

    @Override
    @NotNull
    public Integer getDefaultValue() {
        return this.defaultValue;
    }

    @NotNull
    public final ColorMode getMode() {
        return this.mode;
    }

    public final void setMode(@NotNull ColorMode colorMode) {
        Intrinsics.checkNotNullParameter((Object)colorMode, (String)"<set-?>");
        this.mode = colorMode;
    }

    @Override
    @NotNull
    public Integer getValue() {
        return this.resolveColor();
    }

    @Override
    public void setValue(int newValue) {
        this.mode = new ColorMode.Static(newValue);
        super.setValue(newValue);
    }

    private final int resolveColor() {
        int n;
        ColorMode m = this.mode;
        if (m instanceof ColorMode.Static) {
            n = ((ColorMode.Static)m).getArgb();
        } else if (m instanceof ColorMode.Rainbow) {
            double elapsed = (double)(System.currentTimeMillis() - this.instanceStartTime) / 1000.0;
            float hue = (float)(elapsed * (double)((ColorMode.Rainbow)m).getSpeed() % 1.0 + 1.0) % 1.0f;
            int rgb = Color.HSBtoRGB(hue, ((ColorMode.Rainbow)m).getSaturation(), ((ColorMode.Rainbow)m).getBrightness());
            int alpha = RangesKt.coerceIn((int)((int)(((ColorMode.Rainbow)m).getOpacity() * (float)255)), (int)0, (int)255);
            n = alpha << 24 | rgb & 0xFFFFFF;
        } else if (m instanceof ColorMode.SyncedRainbow) {
            Theme theme = ThemeManager.INSTANCE.getCurrentTheme();
            float hue = ThemeManager.getRainbowHue$default(ThemeManager.INSTANCE, 0.0f, 1, null);
            float sat = theme.getRainbowSaturation();
            float bri = theme.getRainbowBrightness();
            int rgb = Color.HSBtoRGB(hue, sat, bri);
            int alpha = RangesKt.coerceIn((int)((int)(((ColorMode.SyncedRainbow)m).getOpacity() * (float)255)), (int)0, (int)255);
            n = alpha << 24 | rgb & 0xFFFFFF;
        } else if (m instanceof ColorMode.ThemeColor) {
            n = ThemeColorResolver.INSTANCE.resolve(((ColorMode.ThemeColor)m).getPropertyName());
        } else if (m instanceof ColorMode.TweakedTheme) {
            int baseColor = ThemeColorResolver.INSTANCE.resolve(((ColorMode.TweakedTheme)m).getPropertyName());
            n = this.tweakColor(baseColor, ((ColorMode.TweakedTheme)m).getHueOffset(), ((ColorMode.TweakedTheme)m).getSaturationMultiplier(), ((ColorMode.TweakedTheme)m).getBrightnessMultiplier(), ((ColorMode.TweakedTheme)m).getOpacityMultiplier());
        } else {
            throw new NoWhenBranchMatchedException();
        }
        return n;
    }

    private final int tweakColor(int argb, float hueShift, float saturationMult, float brightnessMult, float opacityMult) {
        Color color = new Color(argb, true);
        float[] hsb = Color.RGBtoHSB(color.getRed(), color.getGreen(), color.getBlue(), null);
        float newHue = (hsb[0] + hueShift + 1.0f) % 1.0f;
        float newSat = RangesKt.coerceIn((float)(hsb[1] * saturationMult), (float)0.0f, (float)1.0f);
        float newBri = RangesKt.coerceIn((float)(hsb[2] * brightnessMult), (float)0.0f, (float)1.0f);
        float newAlpha = RangesKt.coerceIn((float)((float)color.getAlpha() / 255.0f * opacityMult), (float)0.0f, (float)1.0f);
        int rgb = Color.HSBtoRGB(newHue, newSat, newBri);
        int alpha = RangesKt.coerceIn((int)((int)(newAlpha * (float)255)), (int)0, (int)255);
        return alpha << 24 | rgb & 0xFFFFFF;
    }

    @Override
    public void read(@NotNull JsonElement element) {
        block23: {
            Object modeType;
            block22: {
                Intrinsics.checkNotNullParameter((Object)element, (String)"element");
                if (!element.isJsonPrimitive()) break block22;
                int argb = element.getAsInt();
                this.mode = new ColorMode.Static(argb);
                super.setValue(argb);
                break block23;
            }
            if (!element.isJsonObject()) break block23;
            JsonObject obj = element.getAsJsonObject();
            Object object = obj.get("mode");
            if (object == null || (object = object.getAsString()) == null) {
                object = "static";
            }
            this.mode = switch (modeType = object) {
                case "static" -> {
                    JsonElement v1 = obj.get("argb");
                    int argb = v1 != null ? v1.getAsInt() : ((Number)super.getValue()).intValue();
                    super.setValue(argb);
                    yield new ColorMode.Static(argb);
                }
                case "rainbow" -> {
                    JsonElement v3 = obj.get("speed");
                    JsonElement v4 = obj.get("saturation");
                    JsonElement v5 = obj.get("brightness");
                    JsonElement v6 = obj.get("opacity");
                    yield new ColorMode.Rainbow(v3 != null ? v3.getAsFloat() : 1.0f, v4 != null ? v4.getAsFloat() : 1.0f, v5 != null ? v5.getAsFloat() : 1.0f, v6 != null ? v6.getAsFloat() : 1.0f);
                }
                case "synced_rainbow" -> {
                    JsonElement v7 = obj.get("speed");
                    JsonElement v8 = obj.get("saturation");
                    JsonElement v9 = obj.get("brightness");
                    JsonElement v10 = obj.get("opacity");
                    yield new ColorMode.SyncedRainbow(v7 != null ? v7.getAsFloat() : 1.0f, v8 != null ? v8.getAsFloat() : 1.0f, v9 != null ? v9.getAsFloat() : 1.0f, v10 != null ? v10.getAsFloat() : 1.0f);
                }
                case "theme" -> {
                    Object v11 = obj.get("propertyName");
                    if (v11 == null || (v11 = v11.getAsString()) == null) {
                        v11 = "accent";
                    }
                    yield new ColorMode.ThemeColor((String)v11);
                }
                case "tweaked_theme" -> {
                    Object v12 = obj.get("propertyName");
                    if (v12 == null || (v12 = v12.getAsString()) == null) {
                        v12 = "accent";
                    }
                    JsonElement v13 = obj.get("hueOffset");
                    JsonElement v14 = obj.get("saturationMultiplier");
                    JsonElement v15 = obj.get("brightnessMultiplier");
                    JsonElement v16 = obj.get("opacityMultiplier");
                    yield new ColorMode.TweakedTheme((String)v12, v13 != null ? v13.getAsFloat() : 0.0f, v14 != null ? v14.getAsFloat() : 1.0f, v15 != null ? v15.getAsFloat() : 1.0f, v16 != null ? v16.getAsFloat() : 1.0f);
                }
                default -> {
                    int argb = ((Number)super.getValue()).intValue();
                    yield new ColorMode.Static(argb);
                }
            };
        }
    }

    @Override
    @NotNull
    public JsonElement write() {
        JsonElement jsonElement;
        ColorMode m = this.mode;
        if (m instanceof ColorMode.Static) {
            jsonElement = (JsonElement)new JsonPrimitive((Number)((ColorMode.Static)m).getArgb());
        } else if (m instanceof ColorMode.Rainbow) {
            JsonObject jsonObject;
            JsonObject $this$write_u24lambda_u240 = jsonObject = new JsonObject();
            boolean bl = false;
            $this$write_u24lambda_u240.addProperty("mode", "rainbow");
            $this$write_u24lambda_u240.addProperty("speed", (Number)Float.valueOf(((ColorMode.Rainbow)m).getSpeed()));
            $this$write_u24lambda_u240.addProperty("saturation", (Number)Float.valueOf(((ColorMode.Rainbow)m).getSaturation()));
            $this$write_u24lambda_u240.addProperty("brightness", (Number)Float.valueOf(((ColorMode.Rainbow)m).getBrightness()));
            $this$write_u24lambda_u240.addProperty("opacity", (Number)Float.valueOf(((ColorMode.Rainbow)m).getOpacity()));
            jsonElement = (JsonElement)jsonObject;
        } else if (m instanceof ColorMode.SyncedRainbow) {
            JsonObject jsonObject;
            JsonObject $this$write_u24lambda_u241 = jsonObject = new JsonObject();
            boolean bl = false;
            $this$write_u24lambda_u241.addProperty("mode", "synced_rainbow");
            $this$write_u24lambda_u241.addProperty("speed", (Number)Float.valueOf(((ColorMode.SyncedRainbow)m).getSpeed()));
            $this$write_u24lambda_u241.addProperty("saturation", (Number)Float.valueOf(((ColorMode.SyncedRainbow)m).getSaturation()));
            $this$write_u24lambda_u241.addProperty("brightness", (Number)Float.valueOf(((ColorMode.SyncedRainbow)m).getBrightness()));
            $this$write_u24lambda_u241.addProperty("opacity", (Number)Float.valueOf(((ColorMode.SyncedRainbow)m).getOpacity()));
            jsonElement = (JsonElement)jsonObject;
        } else if (m instanceof ColorMode.ThemeColor) {
            JsonObject jsonObject;
            JsonObject $this$write_u24lambda_u242 = jsonObject = new JsonObject();
            boolean bl = false;
            $this$write_u24lambda_u242.addProperty("mode", "theme");
            $this$write_u24lambda_u242.addProperty("propertyName", ((ColorMode.ThemeColor)m).getPropertyName());
            jsonElement = (JsonElement)jsonObject;
        } else if (m instanceof ColorMode.TweakedTheme) {
            JsonObject jsonObject;
            JsonObject $this$write_u24lambda_u243 = jsonObject = new JsonObject();
            boolean bl = false;
            $this$write_u24lambda_u243.addProperty("mode", "tweaked_theme");
            $this$write_u24lambda_u243.addProperty("propertyName", ((ColorMode.TweakedTheme)m).getPropertyName());
            $this$write_u24lambda_u243.addProperty("hueOffset", (Number)Float.valueOf(((ColorMode.TweakedTheme)m).getHueOffset()));
            $this$write_u24lambda_u243.addProperty("saturationMultiplier", (Number)Float.valueOf(((ColorMode.TweakedTheme)m).getSaturationMultiplier()));
            $this$write_u24lambda_u243.addProperty("brightnessMultiplier", (Number)Float.valueOf(((ColorMode.TweakedTheme)m).getBrightnessMultiplier()));
            $this$write_u24lambda_u243.addProperty("opacityMultiplier", (Number)Float.valueOf(((ColorMode.TweakedTheme)m).getOpacityMultiplier()));
            jsonElement = (JsonElement)jsonObject;
        } else {
            throw new NoWhenBranchMatchedException();
        }
        return jsonElement;
    }
}

File: src/main/kotlin/org/phantom/api/module/setting/impl/ColorSetting.kt
package org.phantom.api.module.setting.impl

import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonPrimitive
import org.phantom.api.module.setting.Setting
import org.phantom.api.ui.theme.ThemeManager
import java.awt.Color

/** ARGB color picker setting. Value is an ARGB integer (e.g. `0xFFFF0000.toInt()` for red). */
class ColorSetting(
  name: String,
  description: String,
  defaultValue: Int,
  val supportsGradient: Boolean = true,
  val supportsAnimatedGradient: Boolean = true,
) : Setting<Int>(name, description, defaultValue) {

  override val defaultValue: Int = defaultValue

  /** Current color mode (static, rainbow, theme, etc.). */
  var mode: ColorMode = ColorMode.Static(defaultValue)

  /** Instance creation timestamp for per-instance rainbow phase offset. */
  private val instanceStartTime = System.currentTimeMillis()

  /** Dynamically resolved color based on current mode. */
  override var value: Int
    get() = resolveColor()
    set(newValue) {
      mode = ColorMode.Static(newValue)
      super.value = newValue
    }

  /**
   * Resolves the current color based on the active mode.
   * Called on every read of `value` to support dynamic modes (rainbow, theme).
   */
  private fun resolveColor(): Int {
    return when (val m = mode) {
      is ColorMode.Static -> m.argb

      is ColorMode.Rainbow -> {
        // Per-instance rainbow: use instanceStartTime for phase offset
        val elapsed = (System.currentTimeMillis() - instanceStartTime) / 1000.0
        val hue = ((elapsed * m.speed) % 1.0 + 1.0).toFloat() % 1f
        val rgb = Color.HSBtoRGB(hue, m.saturation, m.brightness)
        val alpha = (m.opacity * 255).toInt().coerceIn(0, 255)
        (alpha shl 24) or (rgb and 0x00FFFFFF)
      }

      is ColorMode.SyncedRainbow -> {
        val theme = ThemeManager.currentTheme
        val hue = ThemeManager.getRainbowHue()
        val sat = theme.rainbowSaturation
        val bri = theme.rainbowBrightness
        val rgb = Color.HSBtoRGB(hue, sat, bri)
        val alpha = (m.opacity * 255).toInt().coerceIn(0, 255)
        (alpha shl 24) or (rgb and 0x00FFFFFF)
      }

      is ColorMode.Gradient -> {
        if (m.animated) sampleGradient(m.startArgb, m.endArgb, gradientPhase()) else m.startArgb
      }

      is ColorMode.ThemeColor -> {
        ThemeColorResolver.resolve(m.propertyName)
      }

      is ColorMode.TweakedTheme -> {
        val baseColor = ThemeColorResolver.resolve(m.propertyName)
        tweakColor(baseColor, m.hueOffset, m.saturationMultiplier, m.brightnessMultiplier, m.opacityMultiplier)
      }
    }
  }

  /**
   * Applies HSB/opacity adjustments to a base ARGB color.
   * Used by TweakedTheme mode.
   */
  private fun tweakColor(
    argb: Int,
    hueShift: Float,
    saturationMult: Float,
    brightnessMult: Float,
    opacityMult: Float
  ): Int {
    val color = Color(argb, true)
    val hsb = Color.RGBtoHSB(color.red, color.green, color.blue, null)

    val newHue = (hsb[0] + hueShift + 1f) % 1f
    val newSat = (hsb[1] * saturationMult).coerceIn(0f, 1f)
    val newBri = (hsb[2] * brightnessMult).coerceIn(0f, 1f)
    val newAlpha = ((color.alpha / 255f) * opacityMult).coerceIn(0f, 1f)

    val rgb = Color.HSBtoRGB(newHue, newSat, newBri)
    val alpha = (newAlpha * 255).toInt().coerceIn(0, 255)
    return (alpha shl 24) or (rgb and 0x00FFFFFF)
  }

  override fun read(element: JsonElement) {
    if (element.isJsonPrimitive) {
      // Legacy format: plain ARGB int -> Static mode
      val argb = element.asInt
      mode = ColorMode.Static(argb)
      super.value = argb
    } else if (element.isJsonObject) {
      // New format: JSON object with mode discriminator
      val obj = element.asJsonObject
      val modeType = obj.get("mode")?.asString ?: "static"

      mode = when (modeType) {
        "static" -> {
          val argb = obj.get("argb")?.asInt ?: super.value
          super.value = argb
          ColorMode.Static(argb)
        }

        "rainbow" -> {
          ColorMode.Rainbow(
            speed = obj.get("speed")?.asFloat ?: 1f,
            saturation = obj.get("saturation")?.asFloat ?: 1f,
            brightness = obj.get("brightness")?.asFloat ?: 1f,
            opacity = obj.get("opacity")?.asFloat ?: 1f
          )
        }

        "synced_rainbow" -> {
          ColorMode.SyncedRainbow(
            speed = obj.get("speed")?.asFloat ?: 1f,
            saturation = obj.get("saturation")?.asFloat ?: 1f,
            brightness = obj.get("brightness")?.asFloat ?: 1f,
            opacity = obj.get("opacity")?.asFloat ?: 1f
          )
        }

        "theme" -> {
          ColorMode.ThemeColor(
            propertyName = obj.get("propertyName")?.asString ?: "accent"
          )
        }

        "gradient" -> {
          val startArgb = obj.get("startArgb")?.asInt ?: super.value
          val endArgb = obj.get("endArgb")?.asInt ?: startArgb
          super.value = startArgb
          if (supportsGradient) {
            ColorMode.Gradient(
              startArgb = startArgb,
              endArgb = endArgb,
              direction = obj.get("direction")?.asString ?: ColorMode.DIRECTION_LEFT_TO_RIGHT,
              animated = supportsAnimatedGradient && (obj.get("animated")?.asBoolean ?: false),
            )
          } else {
            ColorMode.Static(startArgb)
          }
        }

        "tweaked_theme" -> {
          ColorMode.TweakedTheme(
            propertyName = obj.get("propertyName")?.asString ?: "accent",
            hueOffset = obj.get("hueOffset")?.asFloat ?: 0f,
            saturationMultiplier = obj.get("saturationMultiplier")?.asFloat ?: 1f,
            brightnessMultiplier = obj.get("brightnessMultiplier")?.asFloat ?: 1f,
            opacityMultiplier = obj.get("opacityMultiplier")?.asFloat ?: 1f
          )
        }

        else -> {
          // Unknown mode: fallback to static
          val argb = super.value
          ColorMode.Static(argb)
        }
      }
    }
  }

  override fun write(): JsonElement {
    return when (val m = mode) {
      is ColorMode.Static -> {
        // Backward compatible: write as plain int
        JsonPrimitive(m.argb)
      }

      is ColorMode.Rainbow -> {
        JsonObject().apply {
          addProperty("mode", "rainbow")
          addProperty("speed", m.speed)
          addProperty("saturation", m.saturation)
          addProperty("brightness", m.brightness)
          addProperty("opacity", m.opacity)
        }
      }

      is ColorMode.SyncedRainbow -> {
        JsonObject().apply {
          addProperty("mode", "synced_rainbow")
          addProperty("speed", m.speed)
          addProperty("saturation", m.saturation)
          addProperty("brightness", m.brightness)
          addProperty("opacity", m.opacity)
        }
      }

      is ColorMode.ThemeColor -> {
        JsonObject().apply {
          addProperty("mode", "theme")
          addProperty("propertyName", m.propertyName)
        }
      }

      is ColorMode.Gradient -> {
        JsonObject().apply {
          addProperty("mode", "gradient")
          addProperty("startArgb", m.startArgb)
          addProperty("endArgb", m.endArgb)
          addProperty("direction", m.direction)
          addProperty("animated", m.animated)
        }
      }

      is ColorMode.TweakedTheme -> {
        JsonObject().apply {
          addProperty("mode", "tweaked_theme")
          addProperty("propertyName", m.propertyName)
          addProperty("hueOffset", m.hueOffset)
          addProperty("saturationMultiplier", m.saturationMultiplier)
          addProperty("brightnessMultiplier", m.brightnessMultiplier)
          addProperty("opacityMultiplier", m.opacityMultiplier)
        }
      }
    }
  }

  private fun gradientPhase(): Float {
    val phase = (System.currentTimeMillis() % 4200L).toFloat() / 4200f
    return ((kotlin.math.cos(phase * Math.PI * 2.0) + 1.0) * 0.5).toFloat()
  }

  private fun sampleGradient(startArgb: Int, endArgb: Int, t: Float): Int {
    val clamped = t.coerceIn(0f, 1f)
    val a = lerpChannel(startArgb ushr 24, endArgb ushr 24, clamped)
    val r = lerpChannel(startArgb ushr 16, endArgb ushr 16, clamped)
    val g = lerpChannel(startArgb ushr 8, endArgb ushr 8, clamped)
    val b = lerpChannel(startArgb, endArgb, clamped)
    return (a shl 24) or (r shl 16) or (g shl 8) or b
  }

  private fun lerpChannel(start: Int, end: Int, t: Float): Int {
    return ((start and 0xFF) + (((end and 0xFF) - (start and 0xFF)) * t)).toInt().coerceIn(0, 255)
  }

}
